# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/WhiteSpaceVisibilityKeeper.cpp
# Commit: fedc2f235876
# Full Hash: fedc2f23587613d83db29c14a37bf38700d69cbf
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2025-07-02 16:04:25
# Description:
#   Bug 1975095 - Make `WhiteSpaceVisibilityKeeper::NormalizeVisibleWhiteSpacesWithoutDeletingInvisibleWhiteSpaces` stop scanning non-white-space char backwards if the white-space sequence ends at the first char r=m_kato
#   
#   If the white-space sequence is only the first character of a `Text`,
#   it scans non-white-space character from `UINT32_MAX`, i.e., it starts
#   scanning it from end of the `Text`.  Then, `startOffset` becomes larger
# ==============================================================================

diff -r 984983f6eb0f -r fedc2f235876 editor/libeditor/WhiteSpaceVisibilityKeeper.cpp
--- a/editor/libeditor/WhiteSpaceVisibilityKeeper.cpp	Wed Jul 02 05:13:04 2025 +0000
+++ b/editor/libeditor/WhiteSpaceVisibilityKeeper.cpp	Wed Jul 02 05:28:34 2025 +0000
@@ -2280,6 +2280,9 @@
     whitespaceOptions += nsTextFragment::WhitespaceOption::NewLineIsSignificant;
   }
   const uint32_t firstOffset = [&]() {
+    if (!*whiteSpaceOffset) {
+      return 0u;
+    }
     const uint32_t offset = textNode.TextFragment().RFindNonWhitespaceChar(
         whitespaceOptions, *whiteSpaceOffset - 1);
     return offset == nsTextFragment::kNotFound ? 0u : offset + 1u;
@@ -2290,6 +2293,7 @@
     return offset == nsTextFragment::kNotFound ? textNode.TextDataLength()
                                                : offset;
   }();
+  MOZ_DIAGNOSTIC_ASSERT(firstOffset <= endOffset);
   nsAutoString normalizedString;
   const char16_t precedingChar =
       !firstOffset ? static_cast<char16_t>(0)
